3.6: Disable new session tickets at runtime

ronald-cron-arm opened the pull request:
## Description
Fix #8749 

## PR checklist
- [x] **changelog** provided
- [ ] **development PR** TODO (content TBD; can wait until after the 3.6.1 release)
- [x] **framework PR** not required
- [x] **3.6 PR** here
- [x] **2.28 PR** not required because: no TLS 1.3
- **tests**  partly provided, partly manual (merge with #9505 and run `programs/ssl/tls13_client_demo.sh`)

gilles-peskine-arm commented on include/mbedtls/ssl.h:
```diff
@@ -4478,6 +4492,20 @@ void mbedtls_ssl_conf_preference_order(mbedtls_ssl_config *conf, int order);
+
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3)
+/**
+ * \brief Enable / Disable TLS 1.3 handling of NewSessionTicket messages (client and TLS 1.3 only).
+ *        (Default: MBEDTLS_SSL_ENABLE_NEW_SESSION_TICKETS_DISABLED)
```
Please document somewhere (here would be a good place) that this defaults off for backward compatibility with non-1.3-aware applications written before 3.6.0. (Similar to e.g. [this](https://github.com/Mbed-TLS/mbedtls/pull/9501/files#diff-b9b23ef0e168bb80241e10e0308a5e7859922a01c0090984b484f2e9f388ce92R1894).)

gilles-peskine-arm replied:
(Repeating an earlier comment so it doesn't get forgotten) Please document somewhere (here would be a good place) that this defaults off for backward compatibility with non-1.3-aware applications written before 3.6.0. (Similar to e.g. [this](https://github.com/Mbed-TLS/mbedtls/pull/9501/files#diff-b9b23ef0e168bb80241e10e0308a5e7859922a01c0090984b484f2e9f388ce92R1894).)

ronald-cron-arm commented on include/mbedtls/ssl.h:
```diff
@@ -4887,6 +4915,13 @@ int mbedtls_ssl_get_session(const mbedtls_ssl_context *ssl,
  * \return         #MBEDTLS_ERR_SSL_HELLO_VERIFY_REQUIRED if DTLS is in use
  *                 and the client did not demonstrate reachability yet - in
  *                 this case you must stop using the context (see below).
+ * \return         #MBEDTLS_ERR_SSL_RECEIVED_NEW_SESSION_TICKET if a TLS 1.3
```
> Please document that after this error, the context remains usable and, after optionally retrieving the ticket data, you may continue using the connection as if this error hadn't happened.

This is done below with the other error codes of the same kind.

mpg replied:
I agree with Ronald that the transient nature of this error is already documented to the same level as the other transient errors.

Regarding Gilles's 2nd point, may I suggest changing the last sentence to "You may either call [...] data, or immediately call this function again with the same parameters if you do not wish to save this ticket now."

gilles-peskine-arm replied:
I'm not going to insist on polishing the documentation in an emergency fix, but “other parts of the documentation suck” is not a good reason to have new documentation that sucks. Especially when it's a surprising change in an LTS, which requires helping users more than usual. But I'm ok with keeping what we have here for this pull request and filing a documentation improvement as a 3.6.2 blocker.

mpg replied:
> “other parts of the documentation suck” is not a good reason to have new documentation that sucks

Agreed, but that's not what I was saying. I think the current structure "return X for success, Y if yyy and then you should do YYY, etc. - anything else is a fatal error and means you must stop using the context" + "warning: if this function returns anything other than X, Y, etc. then you must stop using the context" is clear enough.

If you're unhappy with this structure, OK, we can improve it, but then let's revisit the whole structure, not just this part.

PS: though as I said above, I think being clear on what you should do when receiving this error would be useful. Thinking about it, my previous suggestion was incomplete: "You may either call [...] data _and then call this function again with the same arguments_, or immediately call this function again with the same arguments if you do not wish to save this ticket now." (Emphasis to highlight the part I added, not to be included in the documentation.)

gilles-peskine-arm commented on include/mbedtls/ssl.h:
```diff
@@ -84,8 +84,13 @@
 #define MBEDTLS_ERR_SSL_BAD_CERTIFICATE                   -0x7A00
 /* Error space gap */
 /**
```
Missing: changelog entry

mpg replied:
In particular, for people (if any) who had working code handling 1.3 session tickets with 3.6.0, we must explicitly draw attention to the fact that they need to change their code. (Having to change your working code to keep it working is extremely unexpected in an LTS.)

gilles-peskine-arm commented on include/mbedtls/ssl.h:
```diff
@@ -84,8 +84,13 @@
 #define MBEDTLS_ERR_SSL_BAD_CERTIFICATE                   -0x7A00
 /* Error space gap */
 /**
```
Testing: with the merge of https://github.com/Mbed-TLS/mbedtls/commit/6179f6d456ba94384292cd282cf3359bc3ba1d12 and https://github.com/Mbed-TLS/mbedtls/pull/9505/commits/f0eaaa01c12807fce59832fcc5a1a27bf58aa31b, `programs/ssl/tls13_client_demo.sh` passes.

gilles-peskine-arm replied:
We need a two changelog entries:

* Bug fix: some connections that worked in 3.5 no longer work in 3.6.0.
* Behavior change: applications that were relying on the semi-documented behavior of `MBEDTLS_ERR_SSL_RECEIVED_NEW_SESSION_TICKET` now need to call a function to enable this behavior.

mpg commented on programs/ssl/ssl_client2.c:
```diff
@@ -198,7 +199,8 @@ int main(void)
 #define USAGE_TICKETS                                       \
-    "    tickets=%%d          default: 1 (enabled)\n"
+    "    tickets=%%d              default: 1 (enabled)\n"    \
+    "    new_session_tickets=%%d  default: 1 (enabled)\n"
```
Minor: as a matter of principle, options in `ssl_client2` should default to "library default", that is, not calling the corresponding `conf` function at all. That's the only way to test the the library defaults are what we think they are.

(Not a blocker for the release, as we've agreed on manual testing for now anyway, but wanted to point it out for future reference.)

ronald-cron-arm replied:
Otherwise I would have to enable the new session tickets in all ssl-opt.sh resumption tests. Not changed yet but I can if you think it is worth it.

mpg replied:
Not now, let's leave that to a follow-up.

mpg commented on programs/ssl/ssl_client2.c:
```diff
@@ -1226,6 +1230,11 @@ int main(int argc, char *argv[])
                 goto usage;
             }
+        } else if (strcmp(p, "new_session_tickets") == 0) {
+            opt.tickets = atoi(q);
```
Copy-pasta: I assume you mean `opt.new_session_tickets` here. (This also highlights the lack of testing. Again, not a blocker, but we'll probably want a follow-up.)

ronald-cron-arm replied:
Oops, thanks.

mpg reviewed (commented):
I'm happy with the core of the fix and only have minor feedback.

gilles-peskine-arm commented on include/mbedtls/ssl.h:
```diff
@@ -84,8 +84,13 @@
 /**
- * Received NewSessionTicket Post Handshake Message.
- * This error code is experimental and may be changed or removed without notice.
+ * A TLS 1.3 NewSessionTicket message has been received.
+ * This error code can be returned only on client side if and only if handling
+ * of TLS 1.3 NewSessionTicket messages has been enabled through the
+ * mbedtls_ssl_conf_enable_new_session_tickets() API. This error
+ * code can then be returned by mbedtls_ssl_handshake(),
+ * mbedtls_ssl_handshake_step(), mbedtls_ssl_read(), mbedtls_ssl_write() and
+ * mbedtls_ssl_write_early_data().
```
We can't have long error descriptions because of a [limitation in `generate_errors.pl`: it grabs the whole thing](https://github.com/Mbed-TLS/mbedtls/issues/8783). (Preexisting problem for `MBEDTLS_ERR_SSL_CANNOT_READ_EARLY_DATA` by the way.)

ronald-cron-arm replied:
okay what is the usual way to deal with that then?

gilles-peskine-arm replied:
The usual way is to write a simple sentence as the error description, and lobby our management for time to fix our technical debt.

gilles-peskine-arm commented on include/mbedtls/ssl.h:
```diff
@@ -1447,6 +1455,12 @@ struct mbedtls_ssl_config {
+    /** Whether we handle NewSessionTicket TLS 1.3 messages (<>0) or just ignore them (==0)
+     *  They are ignored by default.
+     */
+    uint8_t MBEDTLS_PRIVATE(new_session_tickets_enabled);
```
BLOCKER: We cannot change the size or layout of `mbedtls_ssl_config` in an LTS branch: it's an ABI break. For example, `mbedtls_ssl_config` might be on the application stack.

We can piggyback on `session_tickets` which currently only uses two values in an 8-bit field.

gilles-peskine-arm commented on library/ssl_msg.c:
```diff
@@ -5595,11 +5595,17 @@ static int ssl_tls13_handle_hs_message_post_handshake(mbedtls_ssl_context *ssl)
 #if defined(MBEDTLS_SSL_SESSION_TICKETS)
             MBEDTLS_SSL_DEBUG_MSG(3, ("NewSessionTicket received"));
-            ssl->keep_current_message = 1;
+            if (ssl->conf->new_session_tickets_enabled ==
+                MBEDTLS_SSL_ENABLE_NEW_SESSION_TICKETS_ENABLED) {
```
Not a blocker, but `xxx_enabled` looks like a boolean setting to me, so I'd expect
```suggestion
            if (ssl->conf->new_session_tickets_enabled) {
```

ronald-cron-arm replied:
`new_session_tickets_enabled` has been removed thus things are different here now.

gilles-peskine-arm commented on include/mbedtls/ssl.h:
```diff
@@ -324,6 +329,9 @@
 #define MBEDTLS_SSL_SESSION_TICKETS_DISABLED     0
 #define MBEDTLS_SSL_SESSION_TICKETS_ENABLED      1
 
+#define MBEDTLS_SSL_ENABLE_NEW_SESSION_TICKETS_DISABLED  0
+#define MBEDTLS_SSL_ENABLE_NEW_SESSION_TICKETS_ENABLED   1
```
This looks like a boolean setting. Why are we defining aliases for true and false? Are we going to add `MBEDTLS_SSL_ENABLE_NEW_SESSION_TICKETS_FILE_NOT_FOUND` later?

ronald-cron-arm replied:
I've just followed what we have for `mbedtls_ssl_conf_session_tickets` and the definition of `MBEDTLS_SSL_SESSION_TICKETS_DISABLED` and `MBEDTLS_SSL_SESSION_TICKETS_ENABLED`.

mpg replied:
At this point we should emulate what other similar settings are doing, which is have the enabled/disabled macros.

(In the future, I'd be happy to `#include <stdbool.h>` and remove all these aliases for true and false.)

gilles-peskine-arm commented on library/ssl_msg.c:
```diff
@@ -5595,11 +5595,17 @@ static int ssl_tls13_handle_hs_message_post_handshake(mbedtls_ssl_context *ssl)
+                                                MBEDTLS_SSL_TLS1_3_NEW_SESSION_TICKET);
+                return MBEDTLS_ERR_SSL_WANT_READ;
+            } else {
+                MBEDTLS_SSL_DEBUG_MSG(3, ("Ignore NewSessionTicket, disabled."));
```
“Ignore” sounds like an instruction for the user, rather than a commentary on what is happening. Also “disabled” written as a second part of the sentence looks like it might mean “I have disabled it” rather than “[because] it is disabled”.
```suggestion
                MBEDTLS_SSL_DEBUG_MSG(3, ("Ignoring NewSessionTicket (signal disabled)."));
```

gilles-peskine-arm commented on tests/src/test_helpers/ssl_helpers.c:
```diff
@@ -2543,6 +2543,9 @@ int mbedtls_test_get_tls13_ticket(
     TEST_EQUAL(ret, 0);
 
+    mbedtls_ssl_conf_enable_new_session_tickets(
+        &client_ep.conf, MBEDTLS_SSL_ENABLE_NEW_SESSION_TICKETS_ENABLED);
```
Test coverage: as of 0e5d4fdfc56540ecc23809f006e83f5cea3c26a5:

* Unit tests have the NewSessionTickets error enabled.
* All the ssl-opt test cases have the NewSessionTickets error enabled.
* `ssl_client1.c` has the NewSessionTickets error disabled, but does not run on the CI.

I'm ok with this as an emergency fix for 3.6.1. For 3.6.2, we must have proper test coverage. At the very least, we must merge https://github.com/Mbed-TLS/mbedtls/pull/9505, which will give us coverage of NewSessionTickets error disabled on the CI. Preferably we should also have unit tests for this setting. It's a setting added to an LTS in a hurry, so good test coverage is especially important to make sure we don't accidentally screw it up more.

ronald-cron-arm replied:
I've added two ssl-opt.sh tests with NewSessionTicket handling disabled at runtime similar to the two ones we have for tickets being disabled at build time.

gilles-peskine-arm commented on include/mbedtls/ssl.h:
```diff
@@ -324,6 +329,9 @@
 #define MBEDTLS_SSL_SESSION_TICKETS_DISABLED     0
 #define MBEDTLS_SSL_SESSION_TICKETS_ENABLED      1
 
+#define MBEDTLS_SSL_ENABLE_NEW_SESSION_TICKETS_DISABLED  0
+#define MBEDTLS_SSL_ENABLE_NEW_SESSION_TICKETS_ENABLED   1
```
Naming: we aren't really enabling “new session tickets”, we're enabling the error code that signals that a NewSessionTicket message was received.

ronald-cron-arm replied:
It is not only enabling the error code, if disabled we do not parse the NewSessionTicket message at all.

gilles-peskine-arm replied:
What does “we do not parse the NewSessionTicket message at all” mean? That we silently discard any message whose header says it's a NewSessionTicket, regardless of whether it's valid or not?

gilles-peskine-arm replied:
For example, if an application calls `mbedtls_ssl_session_save` after each successful call to `mbedtls_ssl_handshake` or `mbedtls_ssl_read`, would it get updated session tickets? I thought it would, since the [code that changes depending on the configuration](https://github.com/Mbed-TLS/mbedtls/pull/9507/files#diff-67127675c5612072a0cb4c12b6dd40b0cfdcc6cdb23839f48ce9e3a2257e77e7) is only code that raises the error, not parsing code.

ronald-cron-arm replied:
discussed offline, I think we have covered that.

gilles-peskine-arm reviewed (changes requested):
* Blocker: ABI break.
* Blocker: changelog entries.
* Should be improved in 3.6.1, must be improved in 3.6.2: documentation.
* Ok for 3.6.1, must be improved in 3.6.2: test coverage.

Manual testing: on the merge of 0e5d4fdfc5 with f0eaaa01c1 (from #9505), `programs/ssl/tls13_client_demo.sh` passes.

ronald-cron-arm commented:
@mpg @gilles-peskine-arm I believe I have addressed all the comments. The CI seems green included the ABI/API checks. Thus this seems in a proper state for further review, thanks.

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -4498,8 +4498,31 @@ void mbedtls_ssl_conf_session_tickets(mbedtls_ssl_config *conf, int use_tickets)
+ *        Contrary to TLS 1.2 tickets, the default value is disabled in Mbed TLS
+ *        3.6.x for backward compatibility with applications based on the Mbed TLS
+ *        client in the default configuration connecting to TLS 1.3 servers
+ *        supporting NewSessionTicket messages.
```
Sorry, I don't understand this sentence, in particular "applications based on the Mbed TLS client".

mpg replied:
Suggestion: "Contrary to TLS 1.2 tickets, the default value is disabled in Mbed TLS 3.6.x for backward compatibility with client applications developed using Mbed TLS 3.5 or earlier with the default configuration." (then perhaps merge with the next paragraph).

ronald-cron-arm replied:
I mean an application that is using the Mbed TLS TLS client like ssl_client2 I'd say.

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -4496,6 +4490,16 @@ void mbedtls_ssl_conf_preference_order(mbedtls_ssl_config *conf, int order);
+ * \return         MBEDTLS_SSL_SESSION_TICKETS_ENABLED or
+ *                 MBEDTLS_SSL_SESSION_TICKETS_DISABLED
+ */
+int mbedtls_ssl_conf_get_session_tickets(const mbedtls_ssl_config *conf);
```
Does this need to be a public function?

ronald-cron-arm replied:
It is used outside ssl_tls.c. It could be a "static inline" function but then I would have to move the definition of  MBEDTLS_SSL_SESSION_TICKETS_TLS1_2/3_BIT/MASK in ssl.h. Can do if you prefer.

mpg replied:
I think we have `ssl_misc.h` for this kind of things (functions used in more than one C file but that don't need to be public).

ronald-cron-arm replied:
Ah yes I did not think about that, That's better, thanks. I will change it.

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -4525,12 +4529,23 @@ void mbedtls_ssl_conf_session_tickets(mbedtls_ssl_config *conf, int use_tickets)
+ * \return         MBEDTLS_SSL_NEW_SESSION_TICKETS_ENABLED or
+ *                 MBEDTLS_SSL_NEW_SESSION_TICKETS_DISABLED
+ */
+int mbedtls_ssl_conf_is_new_session_tickets_enabled(const mbedtls_ssl_config *conf);
```
Does this need to be a public function? Also, naming is inconsistent with `mbedtls_ssl_conf_get_session_tickets()`.

ronald-cron-arm replied:
ditto and yes I know the name is inconsistent but the setting function is named `mbedtls_ssl_conf_enable_new_session_tickets()` and not `mbedtls_ssl_conf_new_session_tickets()`. It is not mbedtls_ssl_conf_new_session_tickets()` because it already exists (setting the number of tickets to send after the handshake on server side). That's why I felt uncomfortable to name it `mbedtls_ssl_conf_get_new_session_tickets()` as I felt it would be associated with `mbedtls_ssl_conf_new_session_tickets()` which is not the case.

mpg replied:
Ah right, makes sense. Anyway, I'll care much less about the name as soon as it's a private function :)

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -5121,9 +5110,16 @@ int mbedtls_ssl_renegotiate(mbedtls_ssl_context *ssl);
+ *                 mbedtls_ssl_conf_enable_new_session_tickets() API. A TLS 1.3
+ *                 NewSessionTicket message has been received and parsed
+ *                 successfully by the client. Ticket data is available in the
+ *                 SSL context and remain available as long as the client does
```
Typo: and _will_ remain

gilles-peskine-arm commented:
Manual testing: on the merge of https://github.com/Mbed-TLS/mbedtls/commit/ with https://github.com/Mbed-TLS/mbedtls/pull/9507/commits/48a9f562ba9a172961af05146039d33e4e14689a (from https://github.com/Mbed-TLS/mbedtls/pull/9505), `programs/ssl/tls*_demo.sh` pass.

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -4465,11 +4465,10 @@ int mbedtls_ssl_conf_max_frag_len(mbedtls_ssl_config *conf, unsigned char mfl_co
- * \brief          Enable / Disable session tickets (client only).
- *                 (Default: MBEDTLS_SSL_SESSION_TICKETS_ENABLED.)
+ * \brief          Enable / Disable TLS 1.2 session tickets (client and TLS 1.2 only).
+ *                 Disabled by default.
```
Should be "Enabled".

ronald-cron-arm replied:
Indeed, thanks.

mpg reviewed (changes requested):
Looks good to me (as far as 3.6.1 is concerned) except for one documentation mistake.

ronald-cron-arm commented on ChangeLog.d/disable-new-session-tickets.txt:
```diff
@@ -0,0 +1,13 @@
+Bugfix
+   * Fix TLS connection abortion in applications using an Mbed TLS client in
```
Yes better thanks.

ronald-cron-arm commented on include/mbedtls/ssl.h:
```diff
@@ -5045,6 +5088,17 @@ int mbedtls_ssl_renegotiate(mbedtls_ssl_context *ssl);
+ *                 mbedtls_ssl_conf_enable_new_session_tickets() API. A TLS 1.3
+ *                 NewSessionTicket message has been received and parsed
+ *                 successfully by the client. Ticket data is available in the
+ *                 SSL context and remain available as long as the client will
```
"since there's no way to control when a message is received"
good point

ronald-cron-arm commented on include/mbedtls/ssl.h:
```diff
@@ -4478,8 +4477,46 @@ void mbedtls_ssl_conf_preference_order(mbedtls_ssl_config *conf, int order);
+/**
+ * \brief Enable / Disable handling of TLS 1.3 NewSessionTicket messages (client and TLS 1.3 only).
+ *
+ *        The handling of TLS 1.3 NewSessionTicket messages is disabled by
+ *        default.
+ *
+ *        Contrary to TLS 1.2 tickets, the default value is disabled in
+ *        Mbed TLS 3.6.x for backward compatibility with client applications
+ *        developed using Mbed TLS 3.5 or earlier with the default
+ *        configuration.
+ *
+ *        Up to Mbed TLS 3.5, in the default configuration TLS 1.3 was
+ *        disabled, and a Mbed TLS client with the default configuration would
+ *        establish a TLS 1.2 connection with a TLS 1.2 and TLS 1.3 capable
+ *        server.
+ *
+ *        Starting with Mbed TLS 3.6.0, TLS 1.3 is enabled by default, and thus
+ *        an Mbed TLS client with the default configuration establishes a
+ *        TLS 1.3 connection with a TLS 1.2 and TLS 1.3 capable server. If
+ *        following the handshake the TLS 1.3 server sends NewSessionTicket
+ *        messages and the Mbed TLS client processes them, this results in
+ *        Mbed TLS high level APIs (mbedtls_ssl_read(),
+ *        mbedtls_ssl_handshake(), ...) to eventually return an
+ *        #MBEDTLS_ERR_SSL_RECEIVED_NEW_SESSION_TICKET non fatal error code
+ *        (see the documentation of mbedtls_ssl_read() for more information on
+ *        that error code). Applications unaware of that TLS 1.3 specific non
+ *        fatal error code are then failing.
```
moved to `mbedtls_ssl_config_defaults()` where we disable the handling of NewSessionTicket messages.

ronald-cron-arm commented on library/ssl_misc.h:
```diff
@@ -2934,8 +2934,37 @@ static inline void mbedtls_ssl_tls13_session_clear_ticket_flags(
+#define MBEDTLS_SSL_SESSION_TICKETS_TLS1_3_MASK \
+    (1 << MBEDTLS_SSL_SESSION_TICKETS_TLS1_3_BIT)
+
+static inline int mbedtls_ssl_conf_get_session_tickets(
```
CI is happy, not changed.

ronald-cron-arm commented:
@gilles-peskine-arm @davidhorstmann-arm I've just pushed three new commits addressing Gilles and Manuel comments. Please have a look.

gilles-peskine-arm reviewed (approved):
LGTM

davidhorstmann-arm reviewed (approved):
LGTM as an incremental review on Manuel's. All feedback seems to be addressed

gilles-peskine-arm commented on ChangeLog.d/disable-new-session-tickets.txt:
```diff
@@ -0,0 +1,14 @@
+     information.
+     Fixes #8749.
+
+Changes
```
This entry should be under “Default behavior changes”.

I only noticed just now and I don't want to go through another round of rework here, so let's fix it during release prep at the changelog polishing step.
